---
title: 'Run GCTA64-COJO analyses'
author: 
- name: "Alejandro M.Carrasco"
  affiliation: UCL
output: 
  html_document:
    code_folding: hide
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---


# Run gcta-cojo on all the coloc eQTL/GWAS SNPs on the DNAJB4 locus conditioning on the DNAJB4 locus lead SNP
```{bash}
--bfile /mnt/rreal/RDS/DATA/AMPPD/AMPPD_w_rsids_rsidsOnly
--chr 1
--maf 0.01
--extract SNPS_OVERLAP.txt
--cojo-p 5e-07
--cojo-file LID_GWAS.ma
--cojo-cond SNPS_CONDITION.txt
--out COJO_LID_CHR1_COND
```


# Run COJO stepwise selection on all the coloc eQTL/GWAS SNPs on the DNAJB4 locus

```{bash}
--bfile /mnt/rreal/RDS/DATA/AMPPD/AMPPD_w_rsids_rsidsOnly
--chr 1
--maf 0.01
--extract SNPS_OVERLAP.txt
--cojo-p 1e-05
--cojo-file LID_GWAS.ma
--cojo-slct
--out COJO_LID_CHR1_SLCT
```


# Run COJO stepwise selection on all chr1 SNPs conditionining on the LRP8 locus GWAS lead SNP

```{bash}
--bfile /mnt/rreal/RDS/DATA/AMPPD/AMPPD_w_rsids_rsidsOnly
--chr 1
--maf 0.01
--cojo-p 1e-05
--cojo-file LID_GWAS.ma
--cojo-cond LEADSNSP_CONDITION.txt
--out COJO_LID_CHR1ALL_LEADSNPSCOND
```


# Conclussion

DNAJB4 rs278853 lead SNP remained nominally associated after running COJO conditioning on the LRP8 lead SNP "rs72673189"
No SNPs other than the lead variant remained significantly associated on the DNAJB4 GWAS-eQTL overlapping SNPs used on the colocalization analysis.
This last conclussion suggest the single causal variant assumption, needed to unbiasedly run coloc test, held


