---
title: 'Power estimation for CPH model'
author: 
- name: "Alejandro M.Carrasco"
  affiliation: UCL
output: 
  html_document:
    code_folding: hide
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---



```{r setup, include=FALSE, warning=FALSE}
library(survSNP)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(data.table)
```


# Description

- Variable features
  - GHR = Genotype hazard ratio
  - rafs = A vector of relative allelic frequencies (for the risk allele)
  - erates = A vector of event rates

- Fixed features 
  - lm = Landmark time used for powering the study
  - pilm = Probability that the time-to-event is greater than lm


# Load the QCed data

```{r}
tpd = fread("../data/PROBAND_ALL.txt")
opdc = fread("../data/OPDC_ALL.txt")
ppmi = fread("../data/PPMI_ALL.txt")
pdbp = fread("../data/PDBP_ALL.txt")
pdstat = fread("../data/PDSTAT_ALL.txt")

```



# Power across GHRs and rafs for erates and ns meta-analysis

Show the current power for a variety of GHRs and rafs for the erates and ns we had in the GWAS meta-analysos

```{r}
GHRs <- seq(1,4,by=1) #HR vector
rafs <- c(0.01,0.02,0.03,0.04,0.05,0.1,0.2,0.3) #AF vector
erates <- 0.15

res<-survSNP.power.table(GHRs,ns = 2687,
                         rafs,
                         erates,
                         pilm=0.5,
                         lm=4.5,
                         model="additive",test="additive",
                         alpha=0.00000005)

head(res[,c("n","GHR","erate","raf","pow0")])
res_trim <- res %>% 
  mutate(power=pow0*100,
         event_rate=erate*100,
         event_rate = as.factor(event_rate),
         GHR = as.factor(GHR),
         raf = as.factor(raf)) %>%
  select(n,GHR,event_rate,raf,power)


current_power = ggplot(res_trim, aes(x = raf,y = power,fill = GHR)) +
  geom_bar(stat="identity",position="dodge") +
  labs(x="Relative AF",
       y="Power (%)") +
  ggtitle("Power to detect genetic associations in cox models") +
  scale_y_continuous(expand = c(0.01,0.01), limits = c(0,100)) +
  scale_x_discrete(expand = c(0.01,0.01)) +
  scale_colour_manual(values = rainbow(4)) +
  theme_bw(base_size = 13) +
  geom_hline(yintercept= 80, linetype='dashed', color=c('black')) +
  theme(panel.spacing = unit (2, "lines"),
        axis.title.x = element_text(margin = margin(t = 20), size = 20),
        axis.title.y = element_text(margin = margin(r = 20), size = 20),
        axis.text.x = element_text(size=20),
        axis.text.y = element_text(size=20),
        plot.title = element_text(hjust = 0.5, size = 25, face = "bold", margin = margin(b = 20)),
        strip.text = element_text(size = 15, face = "bold"),
        legend.text = element_text(size=14),
        legend.title = element_text(size = 15),
        legend.box.margin=margin(0,0,0,50)) 

tiff("FIGURES/CURRENT_POWER.tiff", units="in", width=15, height=15, res=300)
current_power
# insert ggplot code
dev.off()

jpeg("FIGURES/CURRENT_POWER.jpeg", units="in", width=15, height=15, res=500)
current_power
# insert ggplot code
dev.off()
```



# Show power increase as a function of sample size

```{r}


ns<-c(seq(100,41000,by=1000)) #N sequence
#erates<-c(0.0361,0.073,0.127) #event rates 
raf= c(0.005, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1)
GHRs <- c(1.5,seq(2,4,by=1)) 
res <- survSNP.power.table(GHRs,
                           ns,
                           raf,
                           erates = 0.15,
                           pilm=0.5,lm=4.5,
                           model="additive",test="additive",
                           alpha=0.00000005)
#res[,c("n","GHR","erate","raf","pow0")]
res_trim <- res %>% 
  mutate(power=pow0*100,
         event_rate=erate*100,
         event_rate = as.factor(event_rate),
         GHR = as.factor(GHR),
         raf = as.factor(raf)) %>%
  select(n,GHR,event_rate,raf,power)


#Plot

plot_ss_power = res_trim %>%
  dplyr::rename(Relative_AF = raf) %>%
  
  ggplot(aes(x=n, y=power,color=Relative_AF)) +
  geom_line(size = 1) +
  facet_wrap(~GHR, scales = "free") +
  
  labs(x="Sample size (N)",
       y="Power (%)") +
  ggtitle("Power to detect genetic associations in cox models")  +
  scale_colour_manual(values = rainbow(7)) +
  scale_y_continuous(limits = c(0, 100), breaks = seq(0, 100, by = 10), expand = c(0.03,0.03)) +
  scale_x_continuous(limits = c(100, 40000), expand = c(0.01,0.01)) + #, breaks = seq(100, 40000, by = 5700)) +
  
  geom_hline(yintercept=80, linetype="dashed", color = "black") +
  
  theme_bw(base_size = 13) +
  
  theme(panel.spacing = unit (2, "lines"),
        axis.title.x = element_text(margin = margin(t = 20), size = 20),
        axis.title.y = element_text(margin = margin(r = 20), size = 20),
        axis.text.x = element_text(size=20),
        axis.text.y = element_text(size=20),
        plot.title = element_text(hjust = 0.5, size = 25, face = "bold", margin = margin(b = 20)),
        strip.text = element_text(size = 15, face = "bold"),
        legend.text = element_text(size=14),
        legend.title = element_text(size = 15),
        legend.box.margin=margin(0,0,0,50)) 

tiff("FIGURES/POWER_SS.tiff", units="in", width=15, height=15, res=300)
plot_ss_power
# insert ggplot code
dev.off()

jpeg("FIGURES/POWER_SS.jpeg", units="in", width=15, height=15, res=500)
plot_ss_power
# insert ggplot code
dev.off()
```

