---
title: "LID GWAS against PsychEncode datasets"
author: "AlejandroMC, Regina Reynolds"
date: "15/12/2022"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: yes
fontsize: 12pt
---



```{r results="hide", warning=FALSE, message=FALSE}
.libPaths("/data/kronos/kronos/acarrasco/R_libs/")
library(colochelpR)
library(data.table)
library(MafDb.1Kgenomes.phase3.hs37d5)
library(tidyverse)
library(coloc)
library(biomaRt)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(DT)
```

# Load QCed data

```{r}
gwas_qc_mafs = fread("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt")

# Check mafs distribution.
# If maf > 0.5, I am going to flip A1, A2, as well as the effect directionality
gwas_qc_mafs = gwas_qc_mafs %>%
  mutate(maf = ifelse(maf > 0.5, 1-maf, maf),
         beta = ifelse(maf > 0.5, -1*(beta), beta))
```



```{r}
source(here::here("utils", "getgenes.R")) # To import get_genes_within_1Mb_of_signif_SNPs
ensembl_gene_ids_overlapping_1Mb_window_hit =
  get_genes_within_1Mb_of_signif_SNPs(GWAS = gwas_qc_mafs,
                                      pvalue_column="p.value",
                                      CHR_column="CHR",
                                      BP_column="BP",
                                      pval_thresh = 1e-7,
                                      mart=37)

```



# COLOC

## Run coloc with default priors and perform sensitivity check for LRP1B

First we run coloc against eQTLGen

```{r results="hide", warning = FALSE, message = FALSE}
results_path = "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2"
results_path_GWAS_eQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "LID_eQTL_pVal_1e7")
# Path to the GWAS we load
GWAS_path <- "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt"
eQTLgene_path <- "/mnt/rreal/RDS/acarrasco/COLLABS/DATA/eQTLgen/Full_eQTLgen_rsID.MAF.txt"
eQTLgene <- fread(eQTLgene_path)

# df1_N = as.numeric(mean(gwas_qc_mafs$TotalSampleSize))
for (index in seq_along(ensembl_gene_ids_overlapping_1Mb_window_hit)) {

  ensembl_geneID <- ensembl_gene_ids_overlapping_1Mb_window_hit[index]
  print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
  ciseQTL_filtered <- eQTLgene %>%
    filter(!is.na(maf)) %>%
    filter( (maf < 0.5) & (maf > 0)) %>%
    dplyr::filter(Gene == ensembl_geneID) %>%
    check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
    dplyr::filter(!duplicated(SNP))


  if (nrow(ciseQTL_filtered) == 0) {
    next
  }

  df2_N = ceiling(mean(ciseQTL_filtered$NrSamples))

  # Subset from which I Calculate the proportion cases mean.
  df_subset <- gwas_qc_mafs %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP)
  df1_N = ceiling(as.numeric(mean(df_subset$TotalSampleSize)))

  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = gwas_qc_mafs,
                                                           df2 = ciseQTL_filtered,
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval",
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2",
                                                                     "Gene_2"),
                                                        df_1_name = "GWAS",
                                                        df_2_name = "eQTL",
                                                        df1_path = GWAS_path,
                                                        df2_path=eQTLgene_path,
                                                            p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated,
                                 results_dir_path=results_path_GWAS_eQTL)

  }
```



Second, we run coloc against PsychEncoe


```{r results="hide", warning = FALSE, message = FALSE}
results_path = "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2"
results_path_GWAS_eQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "LID_PSYCH_pVal_1e7")
# Path to the GWAS we load
GWAS_path <- "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt"
psychEnc_path <- "/mnt/rreal/RDS/acarrasco/COLLABS/DATA/PsychEncode/Full_hg19_cis-eQTL_rsID_MAF.txt"

psychEncode <- fread(psychEnc_path)


df1_N = as.numeric(unique(gwas_qc_mafs$TotalSampleSize))

for (index in seq_along(ensembl_gene_ids_overlapping_1Mb_window_hit)) {
  ensembl_geneID <- ensembl_gene_ids_overlapping_1Mb_window_hit[index]
  print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
  ciseQTL_filtered <- psychEncode %>%
    filter(!is.na(maf)) %>%
    filter( (maf < 0.5) &  (maf > 0)) %>%
    dplyr::filter(Gene == ensembl_geneID) %>%
    check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
    dplyr::filter(!duplicated(SNP))

  if (nrow(ciseQTL_filtered) == 0) {
    next
  }


  df2_N = ceiling(mean(ciseQTL_filtered$number_of_SNPs_tested, na.rm = T))


  df_subset <- gwas_qc_mafs %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP)
  df1_N = ceiling(as.numeric(mean(df_subset$TotalSampleSize)))


  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = gwas_qc_mafs,
                                                           df2 = ciseQTL_filtered,
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval",
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2",
                                                                     "Gene_2"),
                                                        df_1_name = "GWAS",
                                                        df_2_name = "eQTL",
                                                        df1_path = GWAS_path,
                                                        df2_path=psychEnc_path,
                                                            p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated,
                                 results_dir_path=results_path_GWAS_eQTL)
}

```


# Results

Coloc calculates the posterior probability (PP) for 5 different hypotheses:

- H0: No association with either trait.
- H1: Association with trait 1, not with trait 2.
- H2: Association with trait 2, not with trait 1.
- H3: Association with trait 1 and 2, two independent SNPs.
- H4: Association with trait 1 and trait 2, one shared SNP.

We will set up a PP.H4 > 0.9 to consider the presence of a shared SNP between traits


```{r}
results_eQTL <- merge_coloc_summaries("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/LID_eQTL_pVal_1e7",
                                    add_signif_SNP = F, recursive = T, pattern = ".rda")



  #mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
ensembl <- useEnsembl(biomart = "ensembl",
                     dataset = "hsapiens_gene_ensembl",mirror = "www")

mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 values = results_eQTL$Gene_2,
                 mart = ensembl)

results_eQTL <- results_eQTL %>%
  inner_join(mapping,by = c("Gene_2" = "ensembl_gene_id")) %>%
  rename(HGNC = hgnc_symbol) %>%
  relocate(HGNC, .after = Gene_2)

datatable(results_eQTL, rownames = FALSE,
          options = list(scrollX = TRUE),
           class = 'white-space: nowrap')



results_psychEncode <- merge_coloc_summaries("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/LID_PSYCH_pVal_1e7",
                                 add_signif_SNP = F, recursive = T, pattern = ".rda")

#mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl",mirror = "www")
mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 values = results_psychEncode$Gene_2,
                 mart = ensembl)

results_psychEncode <- results_psychEncode %>%
  inner_join(mapping,by = c("Gene_2" = "ensembl_gene_id")) %>%
  rename(HGNC = hgnc_symbol) %>%
  relocate(HGNC, .after = Gene_2)

results_psychEncode %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```
