---
title: "coloc LID metabrain"
author: "AMCalejandro"
date: "21/9/2022"
output:
  html_document:
    df_print: paged
    highlight: kate
    code_folding: hide
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: yes
---


# Strategy

1. We are going to get all genes within 1Mb from lead SNP in each loci.
Then, we are going to run coloc for each gene using MetaBrain data from cortex tissue

2. We are going to perform a sensitivity analysis using coloc to understand the per hypothessis prior probabilities we get according to the default per SNP priors

3. We will re run coloc adjusting priors so that per hypothesis priors are on a sensible range.


# Set up

## Load libraries and some useful functions


```{r results="hide", warning=FALSE, message=FALSE}
.libPaths("/data/kronos/kronos/acarrasco/R_libs/")
library(colochelpR)
library(data.table)
library(MafDb.1Kgenomes.phase3.hs37d5)
library(tidyverse)
library(coloc)
library(biomaRt)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(DT)
```



## Load data

```{r}
gwas_qc = fread("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_metal_tpd.opdc.ppmi.pdstat.pdbp_BESTMODEL_MOTORBL_stderr_1.tbl")
meta_path = "/mnt/rreal/RDS/DATA/eQTLdata/METABRAIN/METABRAIN_QC/CORTEX/"
```



## Data QC
```{r eval=FALSE}
# Extract maf, reformat
gwas_qc_mafs = gwas_qc %>%
  dplyr::rename(beta= Effect, p.value=`P-value`, se=StdErr,
                maf = Freq1, A1 = Allele1, A2 = Allele2)


# Get rsID to match SNP names from QTL dfs
dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37
gwas_qc_mafs = colochelpR::convert_loc_to_rs(gwas_qc_mafs, dbSNP=dbSNP) %>%
  get_varbeta() %>%
  mutate(GWAS = "LID",
         #CHR = as.numeric(levels(CHR))[CHR])
         CHR = as.numeric(as.character(CHR)),
         BP = as.numeric(BP))

gwas_qc_mafs = gwas_qc_mafs %>%
   filter(!(SNP == "."))

# Deal with multiallelic SNPs
# gwas_qc_mafs <- gwas_qc_mafs %>%
#   dplyr::filter(!( (duplicated(SNP)) |
#                      (duplicated(SNP, fromLast = TRUE)))) %>%
#   dplyr::bind_rows(gwas_qc_mafs %>%
#                      dplyr::filter(duplicated(SNP) | duplicated(SNP, fromLast = TRUE)) %>%
#                      dplyr::group_by(SNP) %>%
#                      dplyr::top_n(1, maf))


#Check mafs distribution.
#If maf > 0.5, I am going to flip A1, A2, as well as the effect directionality
gwas_qc_mafs = gwas_qc_mafs %>%
 mutate(maf = ifelse(maf > 0.5, 1-maf, maf),
        beta = ifelse(maf > 0.5, -1*(beta), beta))



# save
fwrite(gwas_qc_mafs, "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt", quote = F, sep= "\t", col.names=T, row.names=F)
```


```{r include=FALSE}
gwas_qc_mafs = fread("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt")
```



# Getting the genes in +-1MB window

We need to extract all the genes within +/- 1 Mb of all significant hits in the progression to dementia GWAS.

```{r}
source(here::here("utils", "getgenes.R")) # To import get_genes_within_1Mb_of_signif_SNPs
sign_split = gwas_qc_mafs %>%
  filter(p.value < 1e-7)  %>%
  split(f = as.character(.$CHR))


mygenes = sapply(names(sign_split), function(x)
  get_genes_within_1Mb_of_signif_SNPs(GWAS = sign_split[[x]],
                                      pvalue_column="p.value",
                                      CHR_column="CHR",
                                      BP_column="BP",
                                      pval_thresh = 1e-7,
                                      mart=37),
  USE.NAMES = TRUE)
```





# Results

## Results for default priors

```{r}

results_path = "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2"
results_path_GWAS_eQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "LID_METABRAIN_pVal_1e7_PriorsAdjusted")

# Path to the GWAS we load
GWAS_path <- "/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/QC_COLOC.txt"


for (i in seq_along(mygenes)){

  metabrain_path = paste0(meta_path, "QCed_2020-05-26-Cortex-EUR-",
                          names(mygenes)[i],"-biogenformat.txt.gz")

  # Load meta brain and do some QC
  metabrain = fread(metabrain_path)
  metabrain = metabrain %>%
    mutate(ProbeName = gsub("\\..*", "", ProbeName)) %>%
    dplyr::rename(p.value = PValue,
                  maf = MAF)

  # get rsid
  snpname =
    as.data.frame(stringr::str_split_fixed(metabrain$SNPName, n = 4, pattern = ":"))
  metabrain$SNP = snpname$V3


  #Get mean sample size for each locus
  gwas_sign_locus = sign_split[[i]]
  df1_N = ceiling(as.numeric(mean(gwas_sign_locus$TotalSampleSize)))

  # Store genes in chromosome chunk and interate
  genes = mygenes[[i]]

  for (index in seq_along(genes)) {
    ensembl_geneID <- genes[index]
    print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
    metabrain_filtered <- metabrain %>%
      dplyr::filter(ProbeName == ensembl_geneID) %>%
      check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
      dplyr::filter(!duplicated(SNP))

    # get metabrain SS
    df2_N = ceiling(mean(metabrain_filtered$NrSamples))

  # Subset from which I Calculate the proportion cases mean.
  #df_subset <- survDementia_rsID_vbeta %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP)

  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = gwas_qc_mafs,
                                                           df2 = metabrain_filtered,
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval",
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2",
                                                                    "ProbeName_2"),
                                                        df_1_name = "GWAS",
                                                        df_2_name = "eQTL",
                                                        df1_path = GWAS_path,
                                                        df2_path= metabrain_path,
                                                            p12 = 5e-7,
                                                        p1 = 3e-5,
                                                        p2 = 3e-5)
  colochelpR::save_coloc_results(coloc_results_annotated,
                                 results_dir_path=results_path_GWAS_eQTL)
  }
}



```




```{r}
results_eQTL <- merge_coloc_summaries("LID_METABRAIN_pVal_1e7_PriorsAdjusted",
                                    add_signif_SNP = F, recursive = T, pattern = ".rda")

ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl", mirror = "www")


mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                  filters = "ensembl_gene_id",
                  values = results_eQTL$ProbeName_2,
                  mart = ensembl)

 results_eQTL <- results_eQTL %>%
   inner_join(mapping, by = c("ProbeName_2" = "ensembl_gene_id")) %>%
   dplyr::rename(HGNC = hgnc_symbol) %>%
   dplyr::relocate(HGNC, .after = ProbeName_2)

# SHowing the whole table
datatable(results_eQTL, rownames = FALSE,
          options = list(scrollX = TRUE),
           class = 'white-space: nowrap')
```



# Sensitivity analysis for ENSG00000162616 with priors adjusted
```{r}

try(load("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/LID_METABRAIN_pVal_1e7_PriorsAdjusted/LID_eQTL_CORTEX_ENSG00000162616.rda"), silent = F)
try(sensitivity(coloc_results_annotated, rule = "H4 > 0.5"), silent = F)
try(coloc_results_annotated, silent = T)

```



# Sensitivity analysis for ENSG00000143067 with priors adjusted
```{r}

try(load("/mnt/rreal/RDS/acarrasco/ANALYSES_WORKSPACE/GWAS_SURVIVAL/POST_GWAS/V2/LID_METABRAIN_pVal_1e7_PriorsAdjusted/LID_eQTL_CORTEX_ENSG00000143067.rda"), silent = F)
try(sensitivity(coloc_results_annotated, rule = "H4 > 0.5"), silent = F)
try(coloc_results_annotated, silent = T)

```


## Sensitivity analysis



# Session Info
```{r}
utils::sessionInfo()
```